stdnum/gs1_128.py

```diff
@@ -0,0 +1,304 @@
+    '7034': (str, 30, True),
+    '7035': (str, 30, True),
+    '7036': (str, 30, True),
+    '7037': (str, 30, True),
+    '7038': (str, 30, True),
+    '7039': (str, 30, True),
+    '8001': (str, 14, False),
+    '8002': (str, 20, True),
+    '8003': (str, 30, True),
+    '8004': (str, 30, True),
+    '8005': (float, 6, False),
+    '8006': (str, 18, False),
+    '8007': (str, 30, True),
+    '8008': (date, 8, True),
+    '8018': (str, 18, False),
+    '8020': (str, 25, True),
+    '8100': (str, 6, True),
+    '8101': (str, 10, True),
+    '8102': (str, 2, True),
+    '8110': (str, 30, True),
+    '8200': (str, 70, True),
+    '90': (str, 30, True),
+    '91': (str, 90, True),
+    '92': (str, 90, True),
+    '93': (str, 90, True),
+    '94': (str, 90, True),
+    '95': (str, 90, True),
+    '96': (str, 90, True),
+    '97': (str, 90, True),
+    '98': (str, 90, True),
+    '99': (str, 90, True),
+}
+
+
+def _convert_date(val):
+    return datetime.strptime(val, '%y%m%d').date()
+
+
+def _convert_int(val):
+    return int(val)
+
+
+def _convert_float(val):
+    digits = float(_convert_int(val[:2]))
+    value = float(_convert_int(val[2:]))
+    return value / 10 ** digits
+
+
+CONVERTERS = {
+    date: _convert_date,
+    int: _convert_int,
+    float: _convert_float,
+}
+
+
+def _encode_string(val, length):
+    return val.ljust(length)
+
+
+def _encode_date(val, length):
+    return val.strftime('%y%m%d')
+
+
+def _encode_int(val, length):
+    return _encode_string(str(val), length)
+
+
+def _encode_float(val, length):
+    number, decimals = str(val).split('.')
+    digits = str(len(decimals)).rjust(2, '0')
+    number = number.rjust(6 - len(decimals), '0')
+    return _encode_string(digits + number + decimals, length)
+
+
+ENCODERS = {
+    str: _encode_string,
+    date: _encode_date,
+    int: _encode_int,
+    float: _encode_float,
+}
+
+
+def compact(number):
+    """
+    Convert the GS1-128 to the minimal representation.
+
+    This strips the number of any valid separators and removes surrounding
+    whitespace.
+    """
+    return clean(number, '()').strip()
+
+
+def info(code):
+    """
+    Return a dictionary containing the gs1-128 info embeded in code
+
+    Each application code's data is converted to the proper python:
+        * str: for string values
+        * datetime.date: for date values
+        * int: for int values
+        * float: for numeric values
+
+    Escape characters are also removed
+    """
+    code = compact(code)
+    queue = deque(code)
+
+    data = {}
+    identifier = ''
+    while queue:
+        identifier += queue.popleft()
+        if identifier in APPLICATION_CODES:
+            converter, length, variable = APPLICATION_CODES[identifier]
+            value = ''
+            for _ in range(length):
+                try:
+                    char = queue.popleft()
+                    # If it's a stop char do not continue
+                    if variable and char == chr(206):
```

Comment:
I also have a question about this. The problem with `chr(206)` is that is handled differently on Python2 and Python3. The underlying problem is that Code 128 is more or less a binary representation from which a barcode is generated but python-stdnum almost exclusively deals with string (unicode in Python 3).

This means that we need to figure out a text representation of FNC1 in GS1-128. How do other applications that deal with FNC1 or similar codes deal with that? Alternatively, are there common Python libraries available that can generate a bar code from a Code 128 string?

Reply:
The stop characters have [no direct ascii representation](https://stackoverflow.com/questions/31318648/what-is-the-actual-hex-binary-value-of-the-gs1-fnc1-character/31322815#31322815)

We are using barcode scanners and they write they `char(206)` so that's why we proposed to use this caracter but after a second though I think the best option is to make this option configurable.